safekeeper/src/timeline.rs

```diff
@@ -589,6 +592,80 @@ impl Timeline {
         shared_state.get_peers(conf.heartbeat_timeout)
     }
 
+    /// Start streaming tar archive with timeline:
+    /// 1) stream control file under lock;
+    /// 2) hold off WAL removal;
+    /// 3) collect BasebackupContext to understand which WAL segments should be
+    ///    streamed.
+    ///
+    /// Basebackup streams data up to flush_lsn. To make this safe, we must
+    /// check that term doesn't change during the procedure, or we risk sending
+    /// mix of WAL from different histories. Term is remembered in the
+    /// BasebackupContext and checked in finish_basebackup.
+    /// Note that in the last segment some WAL higher than flush_lsn set
+    /// here might be streamed; that's fine as long as terms doesn't change.
+    ///
+    /// In the future we might send only up to commit_lsn to get some valid
+    /// state which later will be recovered by compute, in this case term check
+    /// is not needed.
+    ///
+    /// When returned BasebackupContext is dropped WAL hold is removed.
+    pub async fn start_basebackup<W: AsyncWrite + Unpin + Send>(
+        self: &Arc<Self>,
+        ar: &mut tokio_tar::Builder<W>,
+    ) -> Result<BasebackupContext> {
+        let shared_state = self.read_shared_state().await;
+
+        let cf_path = self.timeline_dir.join(CONTROL_FILE_NAME);
+        let mut cf = File::open(cf_path).await?;
+        ar.append_file(CONTROL_FILE_NAME, &mut cf).await?;
+
+        // We need to stream since the oldest segment someone (s3 or pageserver)
+        // still needs. This duplicates calc_horizon_lsn logic.
+        let from_lsn = min(
+            shared_state.sk.state.remote_consistent_lsn,
```

Comment:
I saw you patched `remote_consistent_lsn` to set it to `timeline_start_lsn` in the other file, but it's probably still possible to have `remote_consistent_lsn` be `0/0` here. I'd added a check here.

Reply:
makes sense, will do

Reply:
done